compiler/qsc_rir/src/passes/ssa_check.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::index_map::IndexMap;
+
+use crate::rir::{BlockId, Instruction, Operand, Program, VariableId};
+
+#[cfg(test)]
+mod tests;
+
+/// Verifies that the program is in Single Static Assignment (SSA) form.
+/// This check ensures that:
+/// - Each variable is assigned exactly once.
+/// - Each variable is used after it is assigned.
+/// - Each variable is used in a block that is dominated by the block in which it is assigned.
+/// - Each phi node references only its predecessors.
+pub fn check_ssa_form(
+    program: &Program,
+    preds: &IndexMap<BlockId, Vec<BlockId>>,
+    doms: &IndexMap<BlockId, BlockId>,
+) {
+    check_phi_nodes(program, preds);
+    let variable_assignments = get_variable_assignments(program);
+    let variable_uses = get_variable_uses(program);
+
+    for (var_id, uses) in variable_uses.iter() {
+        let (def_block_id, def_idx) = variable_assignments
+            .get(var_id)
+            .unwrap_or_else(|| panic!("{var_id:?} is used but not assigned"));
```

Comment:
Would a let...else block be better here?

Reply:
Ah, good call. This went through a couple versions, so I didn't notice I had approached but not reached that pattern.